File: static/heartmates.js
angular
    //div ngapp - HeartmatesApp
    .module('HeartmatesApp', [])

    //managae routes through the routeProvider
    .config(['$routeProvider', function($routeProvider) {
        $routeProvider
            //when for specific cases  - use url and controller
            //loads in ngview
            .when('/', {
                templateUrl: '../static/index.html',
            })
            .when('/checkout', {
                templateUrl: '../static/checkout.html',
//                controller: 'ItemlistController'
            })
            .when('/itemlist', {
                templateUrl: '../static/Whole_Foods.html',
  //              controller: 'ItemlistController'
            })
            .when('/cart', {
                templateUrl: '../static/cart.html',
    //            controller: 'ItemlistController'
            })
            .otherwise({ redirectTo: '/' });
    }])
    .factory('windowAlert', [
        '$window',
        function($window) {
            return $window.alert;
        }
    ])

    .controller('ItemlistController', [
        '$scope', 
        '$http',
        'windowAlert',
        function($scope, $http, windowAlert) {
            $scope.state = {};
            $scope.state.itemsArray = [];
            $scope.state.address = "";
            $scope.state.fee = "";
            $scope.state.created = "";
            $scope.state.eta = "";
            $scope.state.finalList = [
            ];

            $scope.state.itemList = [
            {   name: "Bacon",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "Broccoli",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "Sliced Peaches",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "King Arthur Flour",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "Quaker Steel Cut Oats",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "Extra Virgin Oil",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "Kendall Brooke Salmon",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "Extra Firm Tofu",
                count: 0,
                tempCount: 0,
                score: 0
            },
            {   name: "Prairie Farms Milk",
                count: 0,
                tempCount: 0,
                score: 0
            }];

            console.log("called");

            $scope.increment = function(fooditem) {
                fooditem.tempCount++;
                return;
            };

            $scope.decrement = function(fooditem) {
                fooditem.tempCount--;
                if (fooditem.tempCount < 0) {
                    fooditem.tempCount = 0;
                }
                return;
            };

            $scope.addToCart = function(fooditem) {
                if (fooditem.tempCount != 0) {
                    fooditem.count = fooditem.tempCount;
                    console.log("added " + fooditem.tempCount + " to " + fooditem.name);
                    fooditem.tempCount = 0;
                }

            }

            $scope.firstCheckOut = function() {
                console.log("test");
                for (i = 0; i < $scope.state.itemList.length; i++) {
                    if ($scope.state.itemList[i].count != 0) {
                        $scope.state.itemsArray.push($scope.state.itemList[i].name);
                        console.log('test');
                    }
                }
                console.log($scope.state.itemsArray);

                $http
                    .post('/nutrition_facts/', {
                        items: $scope.state.itemsArray
                    })
                    .success(function(data, status, headers, config) {
                        console.log(data);
                        for (i = 0; i < data.scoreList; i++) {
                             $scope.state.itemList[data.scoreList[i].name] = data.scoreList[i].score;
                             item = {
                                name: data.scoreList[i].name,
                                score: data.scoreList[i].score
                             };
                             $scope.state.finalList.push(score);
                        }
                    })
                    .error(function(data, status, headers, config) {
                });
            };

            $scope.finalCheckOut = function() {
                $http
                    .get('/postmates_delivery/' + $scope.state.address)
                    .success(function(data, status, headers, config) {
                        $scope.state.fee = data.fee;
                        $scope.state.created = data.created;
                        $scope.state.eta = data.eta;
                    })
                    .error(function(data, status, headers, config) {
                    });
            };
        }
    ]);

File: heartmates.py
# PennApps

from flask import Flask, jsonify, render_template, request
import csv
import json
import requests
from requests.auth import HTTPBasicAuth
from threading import Thread

app = Flask(__name__)


@app.route('/')
def home():
    '''This is what you will see if you go to http://127.0.0.1:5000'''
    return render_template('main.html'); 

# @app.route('/nutrition_facts', methods=['POST'])
@app.route('/nutrition_facts/', methods=['POST'])
def nutrition_facts():
    # Get data from post request
    items = json.loads(request.data)['items']
    resps = ['' for x in xrange(len(items))]

    # Set up session
    FE_APIKey = 'n54k3jg9cpkwzvaam4vavc7c'
    params = {'uid': 'user1', 'devid':'device1',
               'appide': 'heartmates', 'f':'json',
               'api_key': FE_APIKey}
    url_create = 'http://api.foodessentials.com/createsession'
    resp = requests.get(url_create, params=params)
    sid = resp.json()['session_id']

    # Set profile
    # API requires entire JSON to be sent to set up profile
    prof_json = '{"session_id":' + str(sid) + ',"nutrients":[{"name":"Calcium","value":"true"},{"name":"Calories","value":"true"},{"name":"Calories from Fat","value":"false"},{"name":"Cholesterol","value":"false"},{"name":"Dietary Fiber","value":"true"},{"name":"Insoluble Fiber","value":"true"},{"name":"Iron","value":"true"},{"name":"Monounsaturated Fat","value":"true"},{"name":"Other Carbohydrate","value":"true"},{"name":"Polyunsaturated Fat","value":"true"},{"name":"Potassium","value":"true"},{"name":"Protein","value":"true"},{"name":"Saturated Fat","value":"false"},{"name":"Saturated Fat Calories","value":"false"},{"name":"Sodium","value":"false"},{"name":"Soluble Fiber","value":"true"},{"name":"Sugar Alcohol","value":"false"},{"name":"Sugars","value":"false"},{"name":"Total Carbohydrate","value":"true"},{"name":"Total Fat","value":"true"},{"name":"Vitamin A","value":"true"},{"name":"Vitamin C","value":"true"}],"allergens":[{"name":"Cereals","value":"false"},{"name":"Coconut","value":"false"},{"name":"Corn","value":"false"},{"name":"Egg","value":"true"},{"name":"Fish","value":"false"},{"name":"Gluten","value":"false"},{"name":"Lactose","value":"true"},{"name":"Milk","value":"true"},{"name":"Peanuts","value":"false"},{"name":"Sesame Seeds","value":"false"},{"name":"Shellfish","value":"false"},{"name":"Soybean","value":"false"},{"name":"Sulfites","value":"false"},{"name":"Tree Nuts","value":"false"},{"name":"Wheat","value":"false"}],"additives":[{"name":"Acidity Regulator","value":"true"},{"name":"Added Sugar","value":"false"},{"name":"Anti-Caking Agents","value":"true"},{"name":"Anti-Foaming Agent","value":"true"},{"name":"Antioxidants","value":"true"},{"name":"Artificial Color","value":"true"},{"name":"Artificial Flavoring Agent","value":"true"},{"name":"Artificial Preservative","value":"true"},{"name":"Bulking Agents","value":"true"},{"name":"Colors","value":"true"},{"name":"Emulsifiers","value":"true"},{"name":"Enzyme","value":"true"},{"name":"Firming Agent","value":"true"},{"name":"Flavor Enhancer","value":"true"},{"name":"Flour Treating Agent","value":"true"},{"name":"Food Acids","value":"true"},{"name":"Gelling Agents","value":"true"},{"name":"Glazing Agent","value":"true"},{"name":"Humectants","value":"true"},{"name":"Leavening Agent","value":"true"},{"name":"Mineral Salt","value":"true"},{"name":"Natural Color","value":"true"},{"name":"Natural Flavoring Agent","value":"true"},{"name":"Natural Preservative","value":"true"},{"name":"Preservatives","value":"true"},{"name":"Propellant","value":"true"},{"name":"Raising Agents","value":"true"},{"name":"Saturated Fat","value":"false"},{"name":"Sequestrant","value":"true"},{"name":"Stabilizers","value":"true"},{"name":"Sweeteners","value":"true"},{"name":"Thickeners","value":"true"},{"name":"Trans Fat","value":"true"},{"name":"Unsaturated Fat","value":"true"},{"name":"Vegetable Gum","value":"true"}],"myingredients":[],"mysort":[{"sort_variable":"Calories","sort_order":1,"variable_type":1}]}'
    params = {'json': prof_json, 'api_key': FE_APIKey}
    url_setprof = 'http://api.foodessentials.com/setprofile'
    resp = requests.post(url_setprof, params=params)


    # Get product info
    url_product = 'http://api.foodessentials.com/productscore'
    # Practice UPC values
    UPC_list = {'Bacon' : '093966004656', 
                'Broccoli': '032601025090',
                'Extra Firm Tofu' : '076371011075',
                'King Arthur Flour' : '030000012031',
                'Prairie Farms Milk' : '041234639642',
                'Quaker Steel Cut Oats' : '021908453361',
                'Sliced Peaches' :'024000167136',
                'Extra Virgin Oil' : '634039000016',
                'Kendall Brooke Salmon' : '15078'}

    # Multithreaded API calls to get item info, fully scalable
    threads = []
    for num in xrange(len(items)):
        params = {'u': UPC_list[items[num]], 'sid':sid, 'f':'json','api_key':FE_APIKey}
        threads.append(Thread(target=get_request, args=(params, num, resps)))
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()

    # Scrape product score from response JSON and return flattened JSON of all product scores
    scores = [str(resp.json()['product']['productscore']) for resp in resps]
    return json.dumps(dict(zip(items,scores)))

def get_request(params, num, resps):
    url_product = 'http://api.foodessentials.com/productscore'
    resps[num] = requests.get(url_product, params=params)


@app.route('/postmates_delivery/<dropoff_address>/')
def postmates_delivery(dropoff_address):

    # Preparing API call
    PM_Test_APIKey = 'd184ecab-5f46-42fd-bbfc-28b73b88cf4e'
    PM_cust_id = 'cus_KAay_YCGWhyi_k'
    url = 'https://api.postmates.com'
    url_delivery = url + '/v1/customers/' + PM_cust_id + '/delivery_quotes'
    headers = {'user': 'd184ecab-5f46-42fd-bbfc-28b73b88cf4e'}
    data = {'pickup_address': '2001 Pennsylvania Avenue Philadelphia, PA 19130',
    'dropoff_address': dropoff_address}
    
    # Sending API request
    resp = requests.post(url_delivery, data=data, auth=HTTPBasicAuth('d184ecab-5f46-42fd-bbfc-28b73b88cf4e', ''))

    # Parsing API response
    rj = resp.json()
    c_vals = (rj['created'].lstrip('0123456789-').lstrip('T').rstrip('Z')).split(':')
    eta_vals = (rj['dropoff_eta'].lstrip('0123456789-').lstrip('T').rstrip('Z')).split(':')
    fee = ('$'+str(int(rj['fee'])/100.0)+'0')[:6]

    # Return fee, request time, and expected delivery time (eta)
    return str(json.dumps({'fee': fee, 'created':str(int(c_vals[0]) - 5) +':'+ c_vals[1] +':'+ c_vals[2], 'eta': str(int(eta_vals[0]) - 5) +':'+ eta_vals[1] +':'+ eta_vals[2]}))





if __name__ == '__main__':
    app.run(debug=True)
